### ISSUE TITLE
syslog-ng: dropping "unnecessary" reload breaks logrotate

### ISSUE THREAD

**Issue by pprindeville** [openwrt/packages#12090]
Maintainer: @BKPepe
Environment: x86_64, generic, master (HEAD)

Description:

I'm looking at:

```
commit 44a16ca3253c1395671905a83d60fa8e604a273a
Author: Karel Kočí <karel.koci@nic.cz>
Date:   Mon Jun 10 09:40:44 2019 +0200

    syslog-ng: improve init and fix service stop
    
    This fixes service stop. Problem was that in default syslog-ng forks to
    background which causes procd to loose track of it. It no longer has in
    such case PID of syslog-ng process and is unable to stop it. This means
    that instance for such process hangs in procd and also it is not
    possible to stop it as daemon. Fix is simple. syslog-ng is now
    instructed to run in foreground.
    
    This commit also drops unnecessary reload section. In default reload
    calls restart and restart in default is implemented as calling stop and
    start. This means that effectively it is implemented same as in case of
    this init reload service implementation.
    
    Signed-off-by: Karel Kočí <karel.koci@nic.cz>
```

and in particular:

```
-reload_service() {
-        stop
-        start
-}
```

and at the same time, the `/etc/logrotate.d/customize` file I wrote which contains:

```
dateext
dateyesterday

# notify syslog-ng to close and reopen files
postrotate
    /etc/init.d/syslog-ng reload
endscript

...
```

where I am *most definitely* using the `reload` functionality and would not all all deem it unnecessary.

I just manually ran the supposedly _equivalent_ default `reload`, and am seeing:

```
root@OpenWrt:~# pidof syslog-ng
8990
root@OpenWrt:~# lsof -p 8990
COMMAND    PID USER   FD      TYPE             DEVICE SIZE/OFF       NODE NAME
syslog-ng 8990 root  cwd       DIR               0,16     3488          5 /
syslog-ng 8990 root  rtd       DIR               0,16     3488          5 /
syslog-ng 8990 root  txt       REG               0,16    16883 4294972408 /usr/sbin/syslog-ng
syslog-ng 8990 root  mem       REG              259,2                5112 /usr/sbin/syslog-ng (path dev=0,16, inode=4294972408)
syslog-ng 8990 root  mem       REG               0,15    16384       3744 /tmp/syslog-ng.persist
syslog-ng 8990 root  mem       REG              259,2                5004 /usr/lib/syslog-ng/libxml.so (path dev=0,16, inode=4294972300)
syslog-ng 8990 root  mem       REG              259,2                5003 /usr/lib/syslog-ng/libtimestamp.so (path dev=0,16, inode=4294972299)
syslog-ng 8990 root  mem       REG              259,2                5002 /usr/lib/syslog-ng/libtfgetent.so (path dev=0,16, inode=4294972298)
syslog-ng 8990 root  mem       REG              259,2                5001 /usr/lib/syslog-ng/libtags-parser.so (path dev=0,16, inode=4294972297)
syslog-ng 8990 root  mem       REG              259,2                5000 /usr/lib/syslog-ng/libsystem-source.so (path dev=0,16, inode=4294972296)
syslog-ng 8990 root  mem       REG              259,2                4999 /usr/lib/syslog-ng/libsyslogformat.so (path dev=0,16, inode=4294972295)
syslog-ng 8990 root  mem       REG              259,2                4998 /usr/lib/syslog-ng/libstardate.so (path dev=0,16, inode=4294972294)
syslog-ng 8990 root  mem       REG              259,2                4997 /usr/lib/syslog-ng/libsnmptrapd-parser.so (path dev=0,16, inode=4294972293)
syslog-ng 8990 root  mem       REG              259,2                4996 /usr/lib/syslog-ng/libpseudofile.so (path dev=0,16, inode=4294972292)
syslog-ng 8990 root  mem       REG              259,2                4995 /usr/lib/syslog-ng/libmap-value-pairs.so (path dev=0,16, inode=4294972291)
syslog-ng 8990 root  mem       REG              259,2                4994 /usr/lib/syslog-ng/liblinux-kmsg-format.so (path dev=0,16, inode=4294972290)
syslog-ng 8990 root  mem       REG              259,2                4993 /usr/lib/syslog-ng/libkvformat.so (path dev=0,16, inode=4294972289)
syslog-ng 8990 root  mem       REG              259,2                1783 /usr/lib/libjson-c.so.4.0.0 (path dev=0,16, inode=4294969079)
syslog-ng 8990 root  mem       REG              259,2                4992 /usr/lib/syslog-ng/libjson-plugin.so (path dev=0,16, inode=4294972288)
syslog-ng 8990 root  mem       REG              259,2                1676 /usr/lib/libcurl.so.4.6.0 (path dev=0,16, inode=4294968972)
syslog-ng 8990 root  mem       REG              259,2                4991 /usr/lib/syslog-ng/libhttp.so (path dev=0,16, inode=4294972287)
syslog-ng 8990 root  mem       REG              259,2                4990 /usr/lib/syslog-ng/libhook-commands.so (path dev=0,16, inode=4294972286)
syslog-ng 8990 root  mem       REG              259,2                4989 /usr/lib/syslog-ng/libgraphite.so (path dev=0,16, inode=4294972285)
syslog-ng 8990 root  mem       REG              259,2                4988 /usr/lib/syslog-ng/libexamples.so (path dev=0,16, inode=4294972284)
syslog-ng 8990 root  mem       REG              259,2                4987 /usr/lib/syslog-ng/libdisk-buffer.so (path dev=0,16, inode=4294972283)
syslog-ng 8990 root  mem       REG              259,2                4986 /usr/lib/syslog-ng/libdbparser.so (path dev=0,16, inode=4294972282)
syslog-ng 8990 root  mem       REG              259,2                4985 /usr/lib/syslog-ng/libcsvparser.so (path dev=0,16, inode=4294972281)
syslog-ng 8990 root  mem       REG              259,2                4984 /usr/lib/syslog-ng/libcryptofuncs.so (path dev=0,16, inode=4294972280)
syslog-ng 8990 root  mem       REG              259,2                4983 /usr/lib/syslog-ng/libconfgen.so (path dev=0,16, inode=4294972279)
syslog-ng 8990 root  mem       REG              259,2                4982 /usr/lib/syslog-ng/libcef.so (path dev=0,16, inode=4294972278)
syslog-ng 8990 root  mem       REG              259,2                4981 /usr/lib/syslog-ng/libbasicfuncs.so (path dev=0,16, inode=4294972277)
syslog-ng 8990 root  mem       REG              259,2                4980 /usr/lib/syslog-ng/libazure-auth-header.so (path dev=0,16, inode=4294972276)
syslog-ng 8990 root  mem       REG              259,2                4978 /usr/lib/syslog-ng/libafuser.so (path dev=0,16, inode=4294972274)
syslog-ng 8990 root  mem       REG              259,2                4977 /usr/lib/syslog-ng/libafstomp.so (path dev=0,16, inode=4294972273)
syslog-ng 8990 root  mem       REG              259,2                4976 /usr/lib/syslog-ng/libafsocket.so (path dev=0,16, inode=4294972272)
syslog-ng 8990 root  mem       REG              259,2                4975 /usr/lib/syslog-ng/libafprog.so (path dev=0,16, inode=4294972271)
syslog-ng 8990 root  mem       REG              259,2                4974 /usr/lib/syslog-ng/libaffile.so (path dev=0,16, inode=4294972270)
syslog-ng 8990 root  mem       REG              259,2                4973 /usr/lib/syslog-ng/libadd-contextual-data.so (path dev=0,16, inode=4294972269)
syslog-ng 8990 root  mem       REG              259,2                4979 /usr/lib/syslog-ng/libappmodel.so (path dev=0,16, inode=4294972275)
syslog-ng 8990 root  mem       REG               0,15       23         87 /tmp/TZ
syslog-ng 8990 root  mem       REG              259,2                1674 /usr/lib/libcrypto.so.1.1 (path dev=0,16, inode=4294968970)
syslog-ng 8990 root  mem       REG              259,2                1893 /usr/lib/libssl.so.1.1 (path dev=0,16, inode=4294969189)
syslog-ng 8990 root  mem       REG              259,2                 886 /lib/libgcc_s.so.1 (path dev=0,16, inode=4294968182)
syslog-ng 8990 root  mem       REG              259,2                1866 /usr/lib/libpcre.so.1.2.11 (path dev=0,16, inode=4294969162)
syslog-ng 8990 root  mem       REG              259,2                1883 /usr/lib/libsecret-storage.so.0.0.0 (path dev=0,16, inode=4294969179)
syslog-ng 8990 root  mem       REG              259,2                1705 /usr/lib/libevtlog-3.26.so.0.0.0 (path dev=0,16, inode=4294969001)
syslog-ng 8990 root  mem       REG              259,2                1734 /usr/lib/libglib-2.0.so.0.5800.3 (path dev=0,16, inode=4294969030)
syslog-ng 8990 root  mem       REG              259,2                1749 /usr/lib/libgthread-2.0.so.0.5800.3 (path dev=0,16, inode=4294969045)
syslog-ng 8990 root  mem       REG              259,2                1737 /usr/lib/libgmodule-2.0.so.0.5800.3 (path dev=0,16, inode=4294969033)
syslog-ng 8990 root  mem       REG              259,2                1900 /usr/lib/libsyslog-ng-3.26.so.0.0.0 (path dev=0,16, inode=4294969196)
syslog-ng 8990 root  mem       REG              259,2                 884 /lib/libc.so (path dev=0,16, inode=4294968180)
syslog-ng 8990 root    0r      CHR                1,3      0t0       5381 /dev/null
syslog-ng 8990 root    1w      CHR                1,3      0t0       5381 /dev/null
syslog-ng 8990 root    2w      CHR                1,3      0t0       5381 /dev/null
syslog-ng 8990 root    3u  a_inode               0,10        0       3086 [eventpoll]
syslog-ng 8990 root    4u  a_inode               0,10        0       3086 [eventfd]
syslog-ng 8990 root    5u  a_inode               0,10        0       3086 [eventfd]
syslog-ng 8990 root    6u  a_inode               0,10        0       3086 [eventfd]
syslog-ng 8990 root    7u  a_inode               0,10        0       3086 [eventfd]
syslog-ng 8990 root    8u  a_inode               0,10        0       3086 [eventfd]
syslog-ng 8990 root    9u  a_inode               0,10        0       3086 [eventfd]
syslog-ng 8990 root   10u      REG               0,15    16384       3744 /tmp/syslog-ng.persist
syslog-ng 8990 root   11u     unix 0x00000000b38146f1      0t0       3781 /dev/log type=DGRAM
syslog-ng 8990 root   12u     IPv6               3783      0t0        UDP localhost:syslog 
syslog-ng 8990 root   13r      REG                0,5        0 4026532013 /proc/kmsg
syslog-ng 8990 root   14u     unix 0x00000000da24b2ac      0t0       3784 /var/syslog-ng.ctl type=STREAM
syslog-ng 8990 root   15w      REG               0,15  8527756       3787 /tmp/log/messages-20200504 (deleted)
syslog-ng 8990 root   18u  a_inode               0,10        0       3086 [timerfd]
root@OpenWrt:~# 
```

Note that it's writing to a deleted file.  Assuming something is equivalent (or unnecessary, for that matter) and actually confirming it to be so are not interchangeable.

```
root@OpenWrt:~# pidof syslog-ng
13623
root@OpenWrt:~# lsof -p 13623
COMMAND     PID USER   FD      TYPE             DEVICE SIZE/OFF       NODE NAME
syslog-ng 13623 root  cwd       DIR               0,16     3488          5 /
syslog-ng 13623 root  rtd       DIR               0,16     3488          5 /
syslog-ng 13623 root  txt       REG               0,16    16883 4294972408 /usr/sbin/syslog-ng
syslog-ng 13623 root  mem       REG              259,2                5112 /usr/sbin/syslog-ng (path dev=0,16, inode=4294972408)
syslog-ng 13623 root  mem       REG               0,15    16384    4568008 /tmp/syslog-ng.persist
syslog-ng 13623 root  mem       REG              259,2                5004 /usr/lib/syslog-ng/libxml.so (path dev=0,16, inode=4294972300)
syslog-ng 13623 root  mem       REG              259,2                5003 /usr/lib/syslog-ng/libtimestamp.so (path dev=0,16, inode=4294972299)
syslog-ng 13623 root  mem       REG              259,2                5002 /usr/lib/syslog-ng/libtfgetent.so (path dev=0,16, inode=4294972298)
syslog-ng 13623 root  mem       REG              259,2                5001 /usr/lib/syslog-ng/libtags-parser.so (path dev=0,16, inode=4294972297)
syslog-ng 13623 root  mem       REG              259,2                5000 /usr/lib/syslog-ng/libsystem-source.so (path dev=0,16, inode=4294972296)
syslog-ng 13623 root  mem       REG              259,2                4999 /usr/lib/syslog-ng/libsyslogformat.so (path dev=0,16, inode=4294972295)
syslog-ng 13623 root  mem       REG              259,2                4998 /usr/lib/syslog-ng/libstardate.so (path dev=0,16, inode=4294972294)
syslog-ng 13623 root  mem       REG              259,2                4997 /usr/lib/syslog-ng/libsnmptrapd-parser.so (path dev=0,16, inode=4294972293)
syslog-ng 13623 root  mem       REG              259,2                4996 /usr/lib/syslog-ng/libpseudofile.so (path dev=0,16, inode=4294972292)
syslog-ng 13623 root  mem       REG              259,2                4995 /usr/lib/syslog-ng/libmap-value-pairs.so (path dev=0,16, inode=4294972291)
syslog-ng 13623 root  mem       REG              259,2                4994 /usr/lib/syslog-ng/liblinux-kmsg-format.so (path dev=0,16, inode=4294972290)
syslog-ng 13623 root  mem       REG              259,2                4993 /usr/lib/syslog-ng/libkvformat.so (path dev=0,16, inode=4294972289)
syslog-ng 13623 root  mem       REG              259,2                1783 /usr/lib/libjson-c.so.4.0.0 (path dev=0,16, inode=4294969079)
syslog-ng 13623 root  mem       REG              259,2                4992 /usr/lib/syslog-ng/libjson-plugin.so (path dev=0,16, inode=4294972288)
syslog-ng 13623 root  mem       REG              259,2                1676 /usr/lib/libcurl.so.4.6.0 (path dev=0,16, inode=4294968972)
syslog-ng 13623 root  mem       REG              259,2                4991 /usr/lib/syslog-ng/libhttp.so (path dev=0,16, inode=4294972287)
syslog-ng 13623 root  mem       REG              259,2                4990 /usr/lib/syslog-ng/libhook-commands.so (path dev=0,16, inode=4294972286)
syslog-ng 13623 root  mem       REG              259,2                4989 /usr/lib/syslog-ng/libgraphite.so (path dev=0,16, inode=4294972285)
syslog-ng 13623 root  mem       REG              259,2                4988 /usr/lib/syslog-ng/libexamples.so (path dev=0,16, inode=4294972284)
syslog-ng 13623 root  mem       REG              259,2                4987 /usr/lib/syslog-ng/libdisk-buffer.so (path dev=0,16, inode=4294972283)
syslog-ng 13623 root  mem       REG              259,2                4986 /usr/lib/syslog-ng/libdbparser.so (path dev=0,16, inode=4294972282)
syslog-ng 13623 root  mem       REG              259,2                4985 /usr/lib/syslog-ng/libcsvparser.so (path dev=0,16, inode=4294972281)
syslog-ng 13623 root  mem       REG              259,2                4984 /usr/lib/syslog-ng/libcryptofuncs.so (path dev=0,16, inode=4294972280)
syslog-ng 13623 root  mem       REG              259,2                4983 /usr/lib/syslog-ng/libconfgen.so (path dev=0,16, inode=4294972279)
syslog-ng 13623 root  mem       REG              259,2                4982 /usr/lib/syslog-ng/libcef.so (path dev=0,16, inode=4294972278)
syslog-ng 13623 root  mem       REG              259,2                4981 /usr/lib/syslog-ng/libbasicfuncs.so (path dev=0,16, inode=4294972277)
syslog-ng 13623 root  mem       REG              259,2                4980 /usr/lib/syslog-ng/libazure-auth-header.so (path dev=0,16, inode=4294972276)
syslog-ng 13623 root  mem       REG              259,2                4978 /usr/lib/syslog-ng/libafuser.so (path dev=0,16, inode=4294972274)
syslog-ng 13623 root  mem       REG              259,2                4977 /usr/lib/syslog-ng/libafstomp.so (path dev=0,16, inode=4294972273)
syslog-ng 13623 root  mem       REG              259,2                4976 /usr/lib/syslog-ng/libafsocket.so (path dev=0,16, inode=4294972272)
syslog-ng 13623 root  mem       REG              259,2                4975 /usr/lib/syslog-ng/libafprog.so (path dev=0,16, inode=4294972271)
syslog-ng 13623 root  mem       REG              259,2                4974 /usr/lib/syslog-ng/libaffile.so (path dev=0,16, inode=4294972270)
syslog-ng 13623 root  mem       REG              259,2                4973 /usr/lib/syslog-ng/libadd-contextual-data.so (path dev=0,16, inode=4294972269)
syslog-ng 13623 root  mem       REG              259,2                4979 /usr/lib/syslog-ng/libappmodel.so (path dev=0,16, inode=4294972275)
syslog-ng 13623 root  mem       REG               0,15       23         87 /tmp/TZ
syslog-ng 13623 root  mem       REG              259,2                1674 /usr/lib/libcrypto.so.1.1 (path dev=0,16, inode=4294968970)
syslog-ng 13623 root  mem       REG              259,2                1893 /usr/lib/libssl.so.1.1 (path dev=0,16, inode=4294969189)
syslog-ng 13623 root  mem       REG              259,2                 886 /lib/libgcc_s.so.1 (path dev=0,16, inode=4294968182)
syslog-ng 13623 root  mem       REG              259,2                1866 /usr/lib/libpcre.so.1.2.11 (path dev=0,16, inode=4294969162)
syslog-ng 13623 root  mem       REG              259,2                1883 /usr/lib/libsecret-storage.so.0.0.0 (path dev=0,16, inode=4294969179)
syslog-ng 13623 root  mem       REG              259,2                1705 /usr/lib/libevtlog-3.26.so.0.0.0 (path dev=0,16, inode=4294969001)
syslog-ng 13623 root  mem       REG              259,2                1734 /usr/lib/libglib-2.0.so.0.5800.3 (path dev=0,16, inode=4294969030)
syslog-ng 13623 root  mem       REG              259,2                1749 /usr/lib/libgthread-2.0.so.0.5800.3 (path dev=0,16, inode=4294969045)
syslog-ng 13623 root  mem       REG              259,2                1737 /usr/lib/libgmodule-2.0.so.0.5800.3 (path dev=0,16, inode=4294969033)
syslog-ng 13623 root  mem       REG              259,2                1900 /usr/lib/libsyslog-ng-3.26.so.0.0.0 (path dev=0,16, inode=4294969196)
syslog-ng 13623 root  mem       REG              259,2                 884 /lib/libc.so (path dev=0,16, inode=4294968180)
syslog-ng 13623 root    0r      CHR                1,3      0t0       5381 /dev/null
syslog-ng 13623 root    1w      CHR                1,3      0t0       5381 /dev/null
syslog-ng 13623 root    2w      CHR                1,3      0t0       5381 /dev/null
syslog-ng 13623 root    3u  a_inode               0,10        0       3086 [eventpoll]
syslog-ng 13623 root    4u  a_inode               0,10        0       3086 [eventfd]
syslog-ng 13623 root    5u  a_inode               0,10        0       3086 [eventfd]
syslog-ng 13623 root    6u  a_inode               0,10        0       3086 [eventfd]
syslog-ng 13623 root    7u  a_inode               0,10        0       3086 [eventfd]
syslog-ng 13623 root    8u  a_inode               0,10        0       3086 [eventfd]
syslog-ng 13623 root    9u  a_inode               0,10        0       3086 [eventfd]
syslog-ng 13623 root   10u      REG               0,15    16384    4568008 /tmp/syslog-ng.persist
syslog-ng 13623 root   11u     unix 0x00000000d5b09a7f      0t0    4568010 /dev/log type=DGRAM
syslog-ng 13623 root   12u     IPv6            4568012      0t0        UDP localhost:syslog 
syslog-ng 13623 root   13r      REG                0,5        0 4026532013 /proc/kmsg
syslog-ng 13623 root   14u     unix 0x000000009759c316      0t0    4568013 /var/syslog-ng.ctl type=STREAM
syslog-ng 13623 root   16w      REG               0,15      884    3220640 /tmp/log/messages
syslog-ng 13623 root   17u  a_inode               0,10        0       3086 [timerfd]
syslog-ng 13623 root   18u  a_inode               0,10        0       3086 [eventpoll]
root@OpenWrt:~# 
```

Note that after an explicit `stop` and `start` it reopens the log file (but this creates a small window of potentially lost logging).

As I left the `init.d` script, it was doing:

```
reload() {
     /usr/sbin/syslog-ng-ctl reload
}
```

and that worked brilliantly with no loss of logging even under heavy load.

Note that `syslog-ng-ctl` talks over a socket to `syslog-ng` and hence doesn't need the PID to communicate with or signal to it.

Please restore this.

**Comment by pprindeville** [comment]
Fixed by PR #12091.

**Comment by pprindeville** [comment]
Resolved.
